### PR TITLE
Update message producer to support PSB v4

### PR DESCRIPTION
**Pull request by codeliner**
- Package + classes are renamed to *message producer
- Classes are adjusted to use new prooph/common message handling
- Coveralls added
- Dependencies updated
- php-cs-fixer added

### CODE REVIEW

**Comment by codeliner** [comment]
ping @prolic

File: src/BernardRouter.php [hunk 1]
```diff
@@ -54,12 +54,14 @@ public function map(Envelope $envelope)
     {
         $message = $envelope->getMessage();
 
-        if (! $message instanceof BernardMessage) throw new \InvalidArgumentException(sprintf(
+        if (! $message instanceof BernardMessage) {
+            throw new \InvalidArgumentException(sprintf(
```
Comment (prolic) [inline src/BernardRouter.php]: Let's use custom exceptions (Prooph\ServiceBus\Message\Bernard\Exception\InvalidArgumentException).

File: src/BernardSerializer.php [hunk 2]
```diff
@@ -33,15 +54,23 @@ public function serialize(Envelope $envelope)
     {
         $message = $envelope->getMessage();
 
-        if (! $message instanceof BernardMessage) throw new \InvalidArgumentException(sprintf(
+        if (! $message instanceof BernardMessage) {
+            throw new \InvalidArgumentException(sprintf(
```
Comment (prolic) [inline src/BernardSerializer.php]: Let's use custom exceptions (Prooph\ServiceBus\Message\Bernard\Exception\InvalidArgumentException).

File: src/MessageProducer.php [hunk 3]
```diff
@@ -1,26 +1,25 @@
...
 
 /**
- * Class MessageDispatcher
+ * Class MessageProducer
  *
  * @package Prooph\ServiceBus\Message\Bernard
  * @author Alexander Miertsch <kontakt@codeliner.ws>
  */
-class MessageDispatcher implements RemoteMessageDispatcher
+class MessageProducer
```
Comment (prolic) [inline src/MessageProducer.php]: Are you sure about removing the interface completely? The interface can also be something like:

```
public function __invoke(Message $message);
```

If you don't want to make the interface a hard dependency, we can do something simliar to what the ZF2's module manager does.

```
if (!$producer instanceof Producer && !is_callable($producer)) {
    throw new Exception\InvalidArgumentException();
}
```

see: https://github.com/zendframework/zend-modulemanager/blob/master/src/Listener/ConfigListener.php#L120

File: src/MessageProducer.php [hunk 4]
```diff
@@ -39,14 +38,15 @@ class MessageDispatcher implements RemoteMessageDispatcher
     public function __construct(Producer $bernardProducer, $queue = null)
     {
         $this->bernardProducer = $bernardProducer;
-        if (! is_null($queue)) $this->useQueue($queue);
+        if (! is_null($queue)) {
```
Comment (prolic) [inline src/MessageProducer.php]: Please change to:

```
null !== $queue
```

consistency is everything ;-)

File: src/MessageProducer.php [hunk 5]
```diff
@@ -57,18 +57,19 @@ public function dispatch(RemoteMessage $message)
      */
     public function useQueue($name)
     {
-        if (! is_string($name)) throw new \InvalidArgumentException("Queue name is not a string!");
+        if (! is_string($name)) {
```
Comment (prolic) [inline src/MessageProducer.php]: same here, null !== $name + custom InvalidArgumentException

File: src/MessageProducer.php (hunk 3 above)
Comment (codeliner) [inline src/MessageProducer.php]: Agreed. The interface doesn't influence the new dispatching logic of prooph/service-bus but make it more explicit how a message producer should look like.

I will add the interface to prooph/service-bus.

File: src/MessageProducer.php (hunk 5 above)
Comment (codeliner) [inline src/MessageProducer.php]: A null check would not be enough. I need to assert **string** here

Comment (prolic) [inline src/MessageProducer.php]: correct!

**Comment by codeliner** [comment]
@prolic the producer lib now throws `Prooph\ServiceBus\Excpetion`s. Do you think that this is enough?

File: src/BernardMessageProducer.php [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        $this->bernardProducer->produce($this->toBernardMessage($message), $this->queue);
+    }
+
+    /**
+     * @param string $name
+     * @throws InvalidArgumentException
+     */
+    public function useQueue($name)
+    {
+        if (! is_string($name)) {
```
Comment (prolic) [inline src/BernardMessageProducer.php]: Assert minLength, or does bernard allow empty strings as queue name?

**Comment by codeliner** [comment]
@prolic Assertion for non empty queue name is added + 100% test coverage :smile:

**Pull request merged**